<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8"/>
		<link href="styles.css" rel="stylesheet"></link>
	</head>
	<body>
	Цитата из книги "Совершенный код" с комментариями. Автор: С.Макконнел
	<p>Купить всю книгу можно на следующих сайтах:
	<ul>
		<li>ozon.ru <a href="http://www.ozon.ru/context/detail/id/138437220/" target="_blank">(открыть в новой вкладке)</a></li>
		<li>market.yandex.ru <a href="https://market.yandex.ru/product--makkonnell-s-sovershennyi-kod-master-klass/1545077" target="_blank">(открыть в новой вкладке)</a></li>
	</ul>
	</p>
	<h3>Часть II. Глава 8.4. Исключения</h3>
<p>Исключения — это специальное средство, позволяющее передать в вызывающий код возникшие ошибки или исключительные ситуации. Если код в некотором методе встречает неожиданную ситуацию и не знает, как ее обработать, то он генерирует исключение, т. е. фактически умывает руки со словами: «Я не знаю, что с этим делать, надеюсь, кто-нибудь другой знает, как на это реагировать!» Код, не имеющий понятия о контексте ошибки, может вернуть управление другой части системы, которая, возможно, лучше знает, как интерпретировать ошибку и сделать с ней что-то осмысленное.</p>
<p>Кроме того, исключения могут быть полезны для упрощения запутанной логики участка кода, как в примере «Переписатьс помощью try-finally» в разделе 17.3. Вот принцип действия исключений: метод, применяя оператор throw, создает объект-исключение. Код какого-либо другого метода, стоящего выше в иерархии вызовов, перехватит это исключение в блоке try-catch.</p>
<p>Популярные языки программирования по-разному реализуют исключения (табл. 8.1):</p>
<p>Табл. 8-1. Поддержка исключений в популярных языках программирования</p>
<table>
	<tr><th>Параметры обработки исключений</th><th>C++</th><th>Java</th><th>Visual Basic</th></tr>
	<tr><td>Поддержка try-catch</td><td>Да</td><td>Да</td><td>Да</td></tr>
	<tr><td>Поддержка try-catch-finally</td><td>Нет</td><td>Да</td><td>Да</td></tr>
	<tr><td>Что генерируется</td><td>Объект класса Exception или производного от него, указатель на объект, объектная ссылка, другие типы данных, например, строка или целое число.</td><td>Объект класса Exception или производного от него.</td><td>Объект класса Exception или производного от него.</td></tr>
	<tr><td>Эффект при не перехваченном исключении</td><td>Вызывается функция std::unexpected(), которая по умолчанию вызывает std::terminate(), в свою очередь по умолчанию вызывающая функцию abort()</td><td>Если это "проверяемое исключение", то прекращается работа потока, в котором ооно возникло. Если это "исключение периода выполнения", то оно игнорируется.</td><td>Программа завершает работу</td></tr>
	<tr><td>Генерируемые исключения должны быть определены в интерфейсе класса</td><td>Нет</td><td>Да</td><td>Нет</td></tr>
	<tr><td>Перехватываемые исключения должны быть определены в интерфейсе класса</td><td>Нет</td><td>Да</td><td>Нет</td></tr>
</table>
<p>Исключения и наследование имеют общее свойство: используемые разумно, они могут уменьшить сложность. Используемые чрезмерно, они могут сделать код абсолютно нечитаемым. Этот раздел содержит предложения по реализации преимуществ исключений и способы избежать трудностей, которые часто с ними связаны.</p>
<p><b>Используйте исключения для оповещения других частей программы об ошибках, которые нельзя игнорировать</b> Основное преимущество исключений состоит в их способности сигнализировать об ошибке так, что ее нельзя проигнорировать (Meyers, 1996). При других подходах к обработке ошибок есть вероятность, что сбойная ситуация останется незамеченной. Исключения устраняют такую возможность.</p>
<p><b>Генерируйте исключения только для действительно исключительных ситуаций</b> Применение исключений должно быть зарезервировано только для действительно исключительных случаев — иначе говоря, для ситуаций, которые нельзя реализовать другими методами кодирования. Исключения используются в таких же обстоятельствах, как и утверждения: для событий, которые не просто редко происходят, а которые никогда не должны случаться.</p>
<p>Исключения представляют собой компромисс между возможностью обработки непредвиденных ситуаций, с одной стороны, и повышением сложности — с другой. Исключения ухудшают инкапсуляцию, требуя от кода, вызывающего метод, знать, какие исключения могут быть сгенерированы внутри него. Это усложняет код, что противоречит Главному Техническому Императиву ПО (см. главу 5), суть которого в снижении сложности.</p>
<p><b>Не используйте исключения по мелочам</b> Если ошибка может быть обработана локально, там ее и обрабатывайте. Не генерируйте в коде неперехватываемое исключение, если ошибка может быть исправлена на месте.</p>
<p><b><span class="comment">При программировании на C++</span> Избегайте генерировать исключения в конструкторах и деструкторах, если только вы не перехватываете их позднее</b> Правила обработки исключений очень быстро усложняются, когда исключения генерируются в конструкторах и деструкторах. Так, в C++ деструктор не вызывается, пока объект не создан полностью. Это значит, что, если код в конструкторе сгенерировал исключение, деструктор вызван не будет, что приведет к возможной утечке ресурсов (Meyers, 1996; Stroustrup, 1997). Аналогичные замысловатые правила относятся и к исключениям внутри деструкторов. Приверженцы языка могут сказать, что запомнить эти правила очень легко. Но все же программисты — простые смертные, у них могут возникнуть трудности с запоминанием правил. Наилучшая программистская практика — избегать излишней сложности, которая создается при написании такого кода.</p>
<p><b>Генерируйте исключения на правильном уровне абстракции</b> Интерфейс метода и класса должен представлять собой целостную абстракцию. Генерируемые исключения — такая же часть интерфейса, как и специальные типы данных. Решив передать исключение, удостоверьтесь, что уровень абстракции исключения
и метода совпадают. Вот как не надо делать:</p>
<p>Плохой пример Java-класса, который генерирует исключение на неверном уровне абстракции</p>
<p class="preformatted">class Employee {
    ...</p>
<p>Объявление исключения с неправильным уровнем абстракции.</p>
<p class="preformatted">    public TaxId GetTaxId() throws EOFException {
        ...
    }
    ...
}
</p>
<p>Функция GetTaxId() передает низкоуровневое исключение EOFException вызывающей стороне. Она не обрабатывает исключение сама, а раскрывает некоторые детали своей реализации, генерируя низкоуровневое исключение. Это привязывает клиентский код не к классу Employee, а к коду внутри класса Employee, генерирующему исключение EOFException. Инкапсуляция нарушена, управляемость кода ухудшается.</p>
<p>Вместо этого код GetTaxId() должен передавать исключение, соответствующее интерфейсу класса, частью которого он является, например, так:</p>
<p>Хороший пример Java-класса, который генерирует исключение на правильном уровне абстракции</p>
<p class="preformatted">class Employee {
    ...</p>
<p>Объявление исключения, соответствующего уровню абстракции.</p>
<p class="preformatted">    public TaxId GetTaxId() throws EmployeeDataNotAvailable {
       ...
    }
    ...
}</p>
<p>Код обработки исключений внутри GetTaxId(), возможно, просто устанавливает соответствие между исключениями io_disk_not_ready и EmployeeDataNotAvailable, что гораздо лучше, так как сохраняется абстракция интерфейса.</p>
<p><b>Вносите в описание исключения всю информацию о его причинах</b> Каждое исключение возникает при определенных обстоятельствах, обнаруженных кодом в момент генерации этого исключения. Эти сведения недоступны тому, кто читает сообщение об исключении. Убедитесь, что это сообщение содержит достаточно информации для понимания причины генерации исключения. Например, если причиной был неправильный индекс элемента массива, включите в описание верхнюю и нижнюю границы массива и некорректное значение индекса.</p>
<p><b>Избегайте пустых блоков catch</b> Иногда возникает искушение оставить без внимания исключение, которое вы не знаете, как обработать. Например:</p>
<p class="preformatted">try {
    ...
    // много кода
    ...
} catch ( AnException exception ) {
}</p>
<p>Такой подход говорит о том, что либо код внутри блока try генерирует исключение без причины, либо код в блоке catch не обрабатывает возможную исключительную ситуацию. Выясните, в чем суть проблемы, и исправьте блоки try или catch.</p>
<p>Изредка можно столкнуться с ситуацией, когда исключение более низкого уровня не соответствует уровню абстракции вызывающего метода. В этом случае хотя бы задокументируйте, почему блок catch должен быть пустым. Это можно сделать в комментариях или записав сообщение в файл журнала, например:</p>
<p>Хороший пример игнорирования исключения на Java</p>
<p class="preformatted">try {
    ...
    // много кода
    ...
} catch ( AnException exception ) {
    LogError( ”Unexpected exception” );
}</p>
<p><b>Выясните, какие исключения генерирует используемая библиотека</b> Если вы работаете с языком, не требующим, чтобы метод или класс объявляли возможные исключения, убедитесь, что вам известно, какие исключения могут возникнуть в коде используемых библиотек. Неперехваченное исключение из библиотеки приведет к аварийному завершению программы так же легко, как и исключение, сгенерированное в вашем коде. Если библиотечные исключения не документированы, создайте код-прототип и протестируйте библиотеки, чтобы их выявить.</p>
<p><b>Рассмотрите вопрос о централизованном выводе информации об исключениях</b> Поддержание целостности в обработке исключений обеспечивает централизованный генератор сообщений об исключениях. Он содержит базу знаний о том, какие это исключения, как каждое из них должно быть обработано, каков формат их сообщений и т. п.</p>
<p>Вот пример упрощенного обработчика исключений. Он просто печатает диагностическое сообщение:</p>
<p>Пример централизованного генератора сообщений об исключениях, часть 1 (Visual Basic)</p>
<p class="preformatted">Sub ReportException( _
ByVal className, _
ByVal thisException As Exception _
)
    Dim message As String
    Dim caption As String
    message = ”Exception: ” & thisException.Message & ”.” & ControlChars.CrLf & _
    ”Class: ” & className & ControlChars.CrLf & _
    ”Routine: ” & thisException.TargetSite.Name & ControlChars.CrLf
    caption = ”Exception”
    MessageBox.Show( message, caption, MessageBoxButtons.OK, _
    MessageBoxIcon.Exclamation )
End Sub</p>
<p>Этот обработчик можно использовать следующим образом:</p>
<p>Пример централизованного генератора сообщений об исключениях, часть 2 (Visual Basic)</p>
<p class="preformatted">Try
    ...
Catch exceptionObject As Exception
    ReportException( CLASS_NAME, exceptionObject )
End Try</p>
<p>Код этой версии ReportException() несложен. В реальных приложениях вы можете сделать отчет настолько кратким или подробным, насколько это необходимо в вашем обработчике исключений.</p>
<p>Если вы решили создать централизованный генератор сообщений, примите во внимание основные проблемы с централизованной обработкой ошибок, обсуждаемые в подразделе «Вызвать процедуру или объект — обработчик ошибок» раздела 8.3.</p>
<p><b>Стандартизуйте использование исключений в вашем проекте</b> Чтобы сохранить процедуру обработки исключений максимально интеллектуально управляемой, вы можете стандартизовать использование исключений несколькими способами.</p>
<ul>
	<li>Если вы работаете с языком, таким как C++, который позволяет генерировать исключения разных типов, стандартизуйте, что конкретно будет создаваться. В целях совместимости с другими языками подумайте об использовании только объектов, порожденных от базового класса Exception.</p>
	<li>Подумайте о создании собственного класса исключений, который может служить базовым классом для всех исключений, возникающих в вашем проекте. Это поможет централизовать и стандартизовать регистрацию, обработку и другие действия с ошибками.</p>
	<li>Определите конкретные случаи, в которых код может использовать синтаксис throw-catch для локальной обработки ошибок.</li>
	<li>Определите конкретные случаи, в которых код может сгенерировать исключение, не перехватываемое локально.</li>
	<li>Решите, будет ли использоваться централизованный генератор сообщений об исключениях.</li>
	<li>Определите, допускаются ли исключения в конструкторах и деструкторах.</li>
</ul>
<p><b>Рассмотрите альтернативы исключениям</b> Некоторые языки поддерживают исключения 5–10 лет и более. Однако до сих пор нет общепринятых правил их безопасного использования.</p>
<p>Некоторые программисты применяют исключения для обработки ошибок только потому, что их язык программирования предоставляет такой механизм. Вам всегда следует принимать во внимание все возможные методы обработки ошибок: локальную обработку ошибок, возврат кода ошибки, запись отладочной информации в файл, прекращение работы системы и др. Обрабатывать ошибки с помощью исключений только потому, что это позволяет язык, — классический пример программирования на языке, а не с использованием языка (см. разделы 4.3 и 34.4).</p>
<p>И напоследок подумайте, действительно ли вашей программе необходимо обрабатывать исключения. Точка. Как заметил Бьерн Страуструп, иногда лучшей реакцией на серьезную ошибку периода выполнения будет освобождение всех ресурсов и прекращение работы. Пусть пользователь перезапустит программу с надлежащими входными данными (Stroustrup, 1997).</p>
	</body>
</html>